### ISSUE TITLE
Log best metric to WandB Sweep

### ISSUE THREAD

**Issue by khituras** [ThilinaRajapakse/simpletransformers#871]
**Is your feature request related to a problem? Please describe.**
During a WandB sweep, the reported metric is updated at each checkpoint. When a run is finished, its final metric will be the last reported one. But the best model of the very same run some steps before could have been much better.

**Describe the solution you'd like**
An option to report the best checkpoint metric at the end of a run would be great.

**Comment by ThilinaRajapakse** [comment]
You could do `eval_model()` with the best checkpoint to report the best metrics, right? I don't think it's a good idea to report the best metric at the end of training as that would make the training charts misleading.

**Comment by khituras** [comment]
Yes, I could - and probably will - do something like this. I think I know what you mean. However in reality, one takes the best checkpoint and not just the final one so I feel I would like to see the best value instead of simple the last. I will be closing this, thanks for the answer!

**Comment by khituras** [comment]
Just for reference, I solved this with closures to avoid another evaluation:

```python
def train():
    """
    This is a closure. We want to keep track of the best F score. But since we pass a function
    to the wandb client, we cannot access variables in outer scope except by using closures.
    """
    bestF=0
    def computeFscore(x,y):
        nonlocal bestF
        p = [round(z) for z in y]
        score = sklearn.metrics.f1_score(x,p)
        recall = sklearn.metrics.recall_score(x, p)
        precision = sklearn.metrics.precision_score(x, p)
        wandb.log({"recall":recall, "precision":precision})
        print(f"Recall: {recall}")
        print(f"Precision: {precision}")
        print(f"F1: {score}")
        if score > bestF:
            bestF = score
        print(f"Current best F: {bestF}")
        return score
    def inner():
        nonlocal bestF
        print("Starting train, initializing WandB.")
        wandb.init(config=model_args)
        model = ClassificationModel(
            architecture,
            modelname,
            use_cuda=True,
            cuda_device=0,
            sweep_config=wandb.config,
            args=model_args
        )
        print("Training model")
        model.train_model(
            train_df, eval_df=eval_df, fscore=computeFscore
        )
        print(f"Finished training. Best F-Score: {bestF}")
        wandb.log({"fscore":bestF})
        bestF = 0
        wandb.finish()
    return inner
```
